Transactions via an internal global session var

juannyG opened the pull request:
An implementation of one of my proposals, resolves #2248  - [Add support for transactions using global sessions](https://github.com/MongoEngine/mongoengine/issues/2248#issuecomment-908504024)

* Maintain state of session when entering and exiting transaction
* Update any collection methods used in mongoengine that accept session
* Update any database methods used in mongoengine that accept session
* Cascades should be accounted for as those methods are wrappers which call the
  same low-level pymongo collection APIs updated above
* Any signals that try to read or write should also be wrapped in the
  transaction

juannyG commented:
Something important for us to keep in mind: [pymongo is extremely upfront that sessions are neither fork-safe nor thread-safe](https://pymongo.readthedocs.io/en/stable/api/pymongo/client_session.html#pymongo.client_session.ClientSession).

At minimum the docs would need to mirror this, but from an implementation stand point, [minor protections could be added here by leveraging python thread's local storage](https://docs.python.org/3/library/threading.html#thread-local-data). That said...it wouldn't actually provide any **real** protections, given that behind the scenes, in a threaded/forked environment, pymongo is still the one in control.

juannyG commented:
Whoops - forgot to give travis the "replicaSet" treatment!

juannyG commented:
OK - I see what's going on - [per the docs](https://docs.mongodb.com/manual/core/transactions/#transactions-api)

> _In MongoDB 4.2 and earlier, you cannot create collections in transactions. Write operations that result in document inserts_
> _(e.g. insert or update operations with upsert: true) must be on existing collections if run inside transactions._
>
> _Starting in MongoDB 4.4, you can create collections in transactions implicitly or explicitly._

So - the test suite is accurate! :smile: I'm testing against a more recent version of Mongo, which is why the tests pass locally.

I think it makes sense to bubble up the exception to a user that they're violating the expectations of the version of Mongo they're using. Should I just have a conditional assertion in my tests that this case should throw an exception for versions<4.4  but >=4.4 the operation is allowed and successful?

juannyG commented:
Digging through the repo, looks like this would be an acceptable approach.

irtzasuhail commented:
is there an estimate on when this might be merged into master?

bagerard commented on mongoengine/queryset/base.py:
```diff
@@ -1715,29 +1728,29 @@ def no_dereference(self):
         map_func = """
-            function() {
-                var path = '{{~%(field)s}}'.split('.');
+            function() {{
```
Can you elaborate on the changes in this method?

juannyG replied:
This wasn't actually something I introduced - I believe it was modified/generated when I had to run the pre-commit hook.

If I removed the changes, it would simply fail...I can double check that on a fresh branch and see what happens...My memory of that is a little vague...but I explicitly remember not having made the change myself but having been forced to by something else.

juannyG replied:
Not sure what happened with that method - but I was able to revert back to what was in the main branch 💪 

resolved!

bagerard commented:
Apologize for the delay, I had a first look at the propasal and it seems to be going in the right direction.
Biggest problem I see right now is that it is not thread safe, using the `run_in_transaction` context manager in 2 different threads would be overwriting the global `_session`, and would be mixing things up badly.

From what I understand the Session is indeed documented as not thread safe, so you can't share a same Session among different threads, but you can let different threads instantiate their own session. So using thread locals should do the trick here. Each thread would create its own Session as needed. We would also need a test that simulates this (I would mix ~ 10 threads, with random delays, some of them saving objects, some of them aborting due to an exception), then assert at the end that only expected items were saved.

bagerard commented on mongoengine/connection.py:
```diff
@@ -398,6 +401,21 @@ def connect(db=None, alias=DEFAULT_CONNECTION_NAME, **kwargs):
 
+def _set_session(session):
+    global _session
+    _session = session
```
I would add a safeguard here and verify that `_session is None` when this is called, otherwise it would mean that we are doing something wrong

bagerard replied:
This would also ensure that people don't start nesting the transaction by doing something like this:
(I'm actually not sure if pymongo would accept to instantiate the second session, let's also add a test for this case so we know what's the behavior) 

```
def do_something():
    with run_in_transaction():
        ...


with run_in_transaction():
    do_something()
```

juannyG replied:
Fun fact - nested transactions are allowed apparently!

I've written two test cases to depict what happens when:
* the nested transactions throws an exc
* the parent transaction throws an exception AFTER the nested transaction was completed.

juannyG replied:
Each time `run_in_transaction` was called - it instantiated a new session object and based on the tests, the rollback states work as I would expect them to.

I guess the question is do we want to force anything at the `mongoengine` level? My vote/opinion is: "if pymongo is OK with it, I'm OK with it"

juannyG replied:
So turns out pymongo does in fact support nested sessions & transactions:
```python
from mongoengine import *
client = connect('mongoenginetest')

A = client.db.A
A.insert_one({"name": "a"})
B = client.db.B
B.insert_one({"name": "b"})


with client.start_session() as s1:
    with s1.start_transaction():
        A.update_one({"name": "a"}, {"$set": {"name": "txA"}}, session=s1)
        with client.start_session() as s2:
            with s2.start_transaction():
                B.update_one({"name": "b"}, {"$set": {"name": "txB"}}, session=s2)
                raise Exception

# Tx was rolled back
list(A.find({})), list(B.find({}))
# ([{'_id': ObjectId('618ae527af50dd19efc2344f'), 'name': 'a'}], [{'_id': ObjectId('618ae527af50dd19efc23450'), 'name': 'b'}])


with client.start_session() as s1:
    with s1.start_transaction():
        A.update_one({"name": "a"}, {"$set": {"name": "txA"}}, session=s1)
        with client.start_session() as s2:
            with s2.start_transaction():
                B.update_one({"name": "b"}, {"$set": {"name": "txB"}}, session=s2)
            raise Exception

# Inner tx was successful, outer tx was rolled back
list(A.find({})), list(B.find({}))
# ([{'_id': ObjectId('618ae527af50dd19efc2344f'), 'name': 'a'}], [{'_id': ObjectId('618ae527af50dd19efc23450'), 'name': 'txB'}])
```

I will start working on expanding the feature to support this. However, do let me know if you prefer to keep things simple for now, for example: "mongoengine will only support one session at a time and will add nested session support in the future"

This is an interesting one :)

juannyG replied:
Turns out the nesting wasn't too tricky!

bagerard commented on tests/test_context_managers.py:
```diff
@@ -412,6 +420,103 @@ def test_query_counter_ignores_particular_queries(self):
+
+        a_doc = A.objects.create(name="a")
+
+        if mongodb_version[0] >= 4 and IS_PYMONGO_GTE_37:
```
Use something similar to @requires_mongodb_gte_44 , checking for mongodb version should be sufficient as pymongo  < 3.7 shouldn't be used with mongodb 4 (https://docs.mongodb.com/drivers/pymongo/#mongodb-compatibility), there was a bug in our CI related with a wrong mapping but it was fixed recently

juannyG replied:
I've modified the tests to use the recommended approach of decorating for skipping.

juannyG commented:
No apologies necessary! Thanks for the feedback - I'll try to address them as soon as I can.

Great idea about thread locals, btw!

juannyG commented on mongoengine/connection.py:
```diff
@@ -234,6 +238,8 @@ def disconnect(alias=DEFAULT_CONNECTION_NAME):
     if alias in _connection_settings:
         del _connection_settings[alias]
 
+    _clear_session()
```
Hmm - I missed this in my update. I'm thinking we'll just want to empty the `_LocalSessions.sessions` property - you're closing a connection, the sessions of that connection won't be of any use to anymore.

juannyG replied:
Then again, given you can disconnect from alias X, but connection Y could still be open. Do you think this now pushes us into the territory of "we need to track sessions across multiple connections"?

juannyG replied:
After thinking about it more and reviewing my original proposal, I do not believe this requires tracking sessions across multiple connections.

Clearing out any remaining sessions should suffice - and I've updated accordingly.

I'm OK **not** supporting someone trying to execute a disconnect of connection X in the middle of a transaction in the context of a session bound to connection Y 

😄

juannyG commented:
OK - I think this is ready for another review round! 💪

solomonr commented:
Would be great to have this reviewed and merged

EVA-08 commented:
Great code, but why not add arguments read concern and write concern? That would be more flexible

juannyG commented:
Thank you for the feedback @EVA-08 - I've added at minimum a placeholder for users to provide a `dict` of kwargs they might need for [start_session](https://pymongo.readthedocs.io/en/stable/api/pymongo/mongo_client.html#pymongo.mongo_client.MongoClient.start_session) or [start_transaction](https://pymongo.readthedocs.io/en/stable/api/pymongo/client_session.html#pymongo.client_session.ClientSession.start_transaction)

We'll see if the maintainers are OK with this.

mclate commented:
Would it be possible to get this PR reviewed? Would be really nice to have transactions available in mongoengine

CC @bagerard

utkarshgupta95 commented:
Hi!! When will this be merged?
Thanks

utkarshgupta95 commented:
Hi!
Just wanted to follow up again on this merge request so that I can plan my project accordingly. Pls let me know if there are any plans to merge it in next few week.
Thanks a lot!!

ShaneHarvey reviewed (changes requested):
Hey @juannyG, PyMongo maintainer here, this is great work! I'm adding a few thoughts on this feature and how it can be improved in the future.

1. threadlocal is good but using contextvars would be better. Since contextvars requires Python 3.7+ we can rewrite it once mongoengine drops support for 3.6.

2. It would be good to add an api to set a session without starting a transaction, that way users can use sessions without transactions. This can be added as a follow on feature. like:
   ```python
   with run_with_session() as session:
       Model1.objects.get()
       Model2.objects.get()
   ```

3. The current implementation will reuse the local session even when using a different connection alias. I believe this is fine since pymongo will raise an error but it would be great to document the limitation so users have a reference point when they encounter this error.

4. I do think we should seriously consider abstracting [session.with_transaction](https://pymongo.readthedocs.io/en/latest/api/pymongo/client_session.html#pymongo.client_session.ClientSession.with_transaction) (added in pymongo 3.9) rather than start_transaction(), otherwise mongoengine users will need to write complex retry logic. We could add this version of `run_in_transaction` and later introduce a second method `run_with_transaction` but I believe it will be confusing to have two apis. The big difference api-wise is that with_transaction accepts a callable instead of being a context manager. I propose we merged this PR and then update the `run_in_transaction` api in a follow PR to use with_transaction, like this:
   ```python
   def run_in_transaction(
       callback, alias=DEFAULT_CONNECTION_NAME, session_kwargs=None, transaction_kwargs=None
   ):
       conn = get_connection(alias)
       session_kwargs = session_kwargs or {}
       with conn.start_session(**session_kwargs) as session:
           transaction_kwargs = transaction_kwargs or {}
           transaction_kwargs["callback"] = callback
           _set_session(session)
           try:
               session.with_transaction(**transaction_kwargs):
           finally:
               _clear_session()
   ```
   
   Then the updated api would be used like this:
   ```python
   def callback():
       a_doc.update(name="a2")
       b_doc.update(name="b2")
   
   run_in_transaction(callback)
   ```

juannyG commented on tests/test_context_managers.py:
```diff
@@ -418,6 +425,302 @@ def test_query_counter_ignores_particular_queries(self):
+                pass
+            except pymongo.errors.OperationFailure as op_failure:
+                """
+                If there's a TransientTransactionError, retry - the lock could not be acquired.
```
Yeah - to be frank, this test was a pain :rofl:

juannyG commented:
Thank you so much for the feedback @ShaneHarvey 

Great call about using `with_transaction`. If you're OK with it, I might just rewrite it now to use `with_transaction`.  Really the "rewrite" will be in the tests to adhere to the new contract. This feels like it might be the right way to kick off transactions in mongoengine - it's cleaner...unlike that unpleasant test I wrote :)

I do see an implication with this approach: if you want to use mongoengine transactions, you have to be using `pymongo>=3.9` - is that OK?

juannyG commented on mongoengine/connection.py:
```diff
@@ -264,6 +267,8 @@ def disconnect(alias=DEFAULT_CONNECTION_NAME):
     if alias in _connection_settings:
         del _connection_settings[alias]
 
+    _local_sessions.clear_all()
```
Yeah...in retrospect I think I was trying to protect from the very weird use case of: someone calls `disconnect` inside of a transaction.

This is definitely over board for that use case. I think, instead, it's OK to say "_don't call disconnect in a transaction - this will result in undefined behavior_"

juannyG commented:
I believe I've addressed the feedback provided, but do let me know if anything else comes to mind.

I'm more than happy to do a follow up PR modifying the API to use `with_transaction` as I completely agree that it is going to be significantly more friendly.

My only concern with doing it as a follow up is that it would be a non-backwards compatible change and the the implication I mentioned above (`pymongo>=3.9` is a requirement).

Looking forward to your thoughts!

idoshr commented:
This PR https://github.com/MongoEngine/mongoengine/pull/2768
Fix the issue of transaction session and transaction between DB

ShaneHarvey commented:
> If you're OK with it, I might just rewrite it now to use with_transaction. Really the "rewrite" will be in the tests to adhere to the new contract. This feels like it might be the right way to kick off transactions in mongoengine - it's cleaner...unlike that unpleasant test I wrote :)

Yes I totally agree. 

> I do see an implication with this approach: if you want to use mongoengine transactions, you have to be using pymongo>=3.9 - is that OK?

This is okay. You could even check the pymongo version at runtime ([like mongoengine does here](https://github.com/MongoEngine/mongoengine/blob/a0ed25faf86e9424dd9982033f44785c06d29d00/mongoengine/pymongo_support.py#L8C1-L8C16)) and give a nice error message if you like.

juannyG commented:
I've completed the modifications so that we now abstract `with_transaction` as opposed to `start_transaction`. If `run_in_transaction` is called and `pymongo<3.9` is being used, we raise a mongoengine `OperationError` and unit tests are skipped when `pymongo<3.9` is active.

The other additional change I made was to move `run_in_transaction` out of `context_managers.py` because it's no longer a context manager. Instead, I put it closer to the newly added `session` stuff in `connection.py` (and updated the tests accordingly, as well).

Let me know if anything else come to mind!

ShaneHarvey commented on tests/test_connection.py:
```diff
@@ -604,7 +618,16 @@ def test_read_preference_from_parse(self):
+            is now how mongod is run for the test suites for transactions.
+            """
+            connect(
+                "mongoenginetest2", alias="t2", host="127.0.0.1", directConnection=True
```
Is this change still needed? It seems unrelated to the run_in_transaction feature.

juannyG replied:
For this test I needed to make a change in order for it to pass. It seemed to me that we needed to account for how pymongo deals with connections to replica sets (which is needed for transactions) as of >= 4.0 https://pymongo.readthedocs.io/en/stable/migrate-to-pymongo4.html#mongoclient

ShaneHarvey replied:
Oh thanks for the explanation. The issue is that without directConnection=True, the client will discover the real server address of the replica set member so `mongo_connections["t2"].address[0]` is equal to `localhost` not `127.0.0.1`. This seems fine, although the test could be rewritten to just do this instead of checking the address property:
```python
assert mongo_connections["t1"] is not mongo_connections["t2"]
```

juannyG replied:
:100: great call. I went ahead and made this update as well

ShaneHarvey reviewed (approved):
This looks good to me! Thanks for all the effort @juannyG!

@bagerard what do you think?

bagerard commented:
again apologize for the delay, I'm finally getting back to this.

@ShaneHarvey I find the "callback api" (i.e `session.with_transaction(callback)`) really awkward, this is typically what context managers are for...

I agree that the retry logic is great to abstract from end users but Pymongo should provide the most flexible and pythonic approach. Using callback and lambda's also interfer with mypy so unless I'm missing something it really provides no advantages...

In fact if Pymongo would offer the "context manager API" (that would include the retry logic), we could easily write a "callback" wrapper in just a few lines of Python but we can't do the opposite and derive a context manager out of the "callback api" which is a pity...

juannyG commented:
@bagerard Thank you for the feedback.

Would you prefer [the implementation I had](https://github.com/MongoEngine/mongoengine/compare/master...juannyG:mongoengine:4a7d24ab23fbcf08beeabec308211a6f029e918f) prior to the callback modification - or is there something else you have in mind?

ShaneHarvey commented:
PyMongo _does_ have a context manager API but the drawback is that is not possible to have retries with such an API [as mentioned above](https://github.com/MongoEngine/mongoengine/pull/2569#pullrequestreview-1548484351). Retries are not possible because the context block can only ever be run once; a context manager can not re-run the block it contains.

bagerard commented:
I missed that it was actually re-calling the method :grimacing: , I understand now, thanks for (re)-clarifying.

In practice when you just chain a bunch of pure MongoDB operations, using the callback thing is a perfect fit. When used in a more involved application, like a complex web application (which is what we are building on top of MongoEngine at work), I don't think it would be a sane default to just retry a chain of calls, simply failing the overall transaction and API call sounds like a much much safer thing to do (otherwise you must be careful that first execution is not going to leave stuff behind that could interfer with the second execution). Other obvious issue with this automatic re-call, is that pymongo is silently doing that, without even logging the error or the fact that it's retrying :grimacing: , again makes sense for a chain of pure db operations, but a nightmare to debug in a more complex setup.

Long story short, I think it's a must to provide a simple context manager API for opening the transaction in MongoEngine, without any default retry logic. Next to that we can also support the "callback-api" but it's a nice to have IMO.

ShaneHarvey commented:
> I don't think it would be a sane default to just retry a chain of calls

I understand this concern but I don't think this is as big of a concern as you say since the primary use case for transaction is a bunch of pure MongoDB operations. Doing other work inside the transaction extends its lifespan and increases the chance of write conflicts. Also, basically all users will need to write retry logic so the same problem still exists in either API.

> Other obvious issue with this automatic re-call, is that pymongo is silently doing that, without even logging the error or the fact that it's retrying 😬 , again makes sense for a chain of pure db operations, but a nightmare to debug in a more complex setup.

We are actually working on adding logging into PyMongo right now: https://jira.mongodb.org/browse/PYTHON-3113

The huge benefit to using the callback approach is that it makes applications much more resilient to transient errors (which are very common when using transactions). Users will be much less likely to have to write custom and error prone transaction retry logic. By opting for a context api you'll end up with a lot of users scratching their heads diagnosing transient write conflict errors.

I would say that if we do end up adding two APIs, then the callback based one should still be the recommended approach in the docs and examples. The context based one could be a "power user" feature for those who want to customize their own retry behavior.

ShaneHarvey commented on mongoengine/connection.py:
```diff
@@ -463,3 +467,89 @@ def connect(db=None, alias=DEFAULT_CONNECTION_NAME, **kwargs):
+        transaction_kwargs["callback"] = callback
+        _set_session(session)
+        try:
+            session.with_transaction(**transaction_kwargs)
```
PyMongo's with_transaction returns the callback's value so I believe this should be `return session.with_transaction(...)`. Otherwise it becomes unwieldy to return a value.

ShaneHarvey commented:
One other idea: if you're very concerned about the retry behavior, one alternative is to expose a configurable knob like this:
```python
run_in_transaction(callback, retry_enabled=False/True)
```

That would at least let you expose a single txn api instead of 2 different ones. I'm ambivalent but it's worth a thought.

bagerard commented:
I'm expecting MongoEngine to be used mostly in web applications and users will be opening their transactions quite early, wrapping all their business logic. In the middle of the chain of calls we may have disk operations, network calls, use of globals/threads/threadlocals, stateful objects, etc. Safe repetition of call is not something you get out of the box so I can't recommend that setup in a general purpose ORM like MongoEngine.

That being said, I don't have much experience with transactions in Mongo, is there anything in MongoDB transactions that is different from SQL transactions and that would make them fail more often? We don't see such callback api with repeated calls for dealing with transactions in extremely popular framework like SQLAlchemy or Django, how do you explain this?

elormkoto commented:
Hi yall. is there any update on this MR? Is there anything I can do to help push this along?

juannyG commented:
> Hi yall. is there any update on this MR? Is there anything I can do to help push this along?

I was waiting to see if there would be a response to @bagerard 's last question - it seemed to me that we were still waiting on alignment. All that said: @bagerard to keep things moving, I can [roll this back to a state](https://github.com/MongoEngine/mongoengine/compare/master...juannyG:mongoengine:4a7d24ab23fbcf08beeabec308211a6f029e918f#diff-571b6b83cf005160ade95ce6611aaeca333eb51355ced6fd2013d74db474ae50R304) that I think reflects more of what you were expecting wrt "_simple context manager without retry logic_". More than happy to get that done.

The test cases also include [examples of how a user would have to deal with retrying transaction failures](https://github.com/MongoEngine/mongoengine/compare/master...juannyG:mongoengine:4a7d24ab23fbcf08beeabec308211a6f029e918f#diff-79d04ecfcff29071fa7640125977132e6f8dd243d26a13f6de9e3a8736fde3dfR675).

scornz commented:
More than happy to hop on and help out with this MR, anything else that needs to get done in order for this to get merged?

juannyG commented:
@bagerard - I have rolled the transactions implementation back to a state where it is simply a context manager which is leaning on pymongo's `start_transaction` context manager. If there is any issue with the transaction within pymongo or mongoDB, it will be up to the application to decide how to handle it. I hope this aligns more with your expectations.

bagerard commented:
Looks good to me, I'm doing some polishing on top of it and should be able to merge this later this week. This big change will also be a good reason to switch from 0.x to 1.x versioning :rocket:

nickphbrown commented:
Hey @bagerard, I'm interested in this change, can I check what the plan is in regards to releasing it?

mcsimps2 commented:
Regarding nested transaction behavior, the standard behavior I'm used to seeing in other database systems is that
- If the child transaction fails, both the child and parent transactions are rolled back
- If the child commits but the parent fails afterwards, both are rolled back

Our implementation in this PR deviates from what developers might expect from the second bullet point in that, if the child commits but the parent fails afterwards, the child is not rolled back.  The child commits while the parent roll backs, which can lead to weird state discrepancies, especially if the operation is retried.

My team at work has altered the implementation in our fork to check `in_transaction` before creating another session so that nested transactions behave as we would expect.  (We also added a flag to enable/disable retries based on the conversation above - it's off by default since it requires an idempotent execution in the event of a retry.) Eager to hear what others think of this.

ShaneHarvey commented:
I would expect MongoEngine to error if a nested transaction is attempted because MongoDB does not support nested transactions.

mcsimps2 commented:
Since there is no actual support for real nested transactions in MongoDB, I'm okay with "appending" to the current transaction if we happen to encounter a nested transaction in the code - that is at least desirable behavior in my codebase as we have some functions that are re-usable but also standalone.  Meaning they may be called as a standalone transaction or as a nested transaction in part of a bigger transaction.

One problem with the current approach is that write conflicts are pretty much guaranteed if a transaction and a nested transaction both modify the same document.  This is because the current approach treats these as completely separate transactions rather than acknowledging the nesting relationship between them.

Here's an example of a write conflict in the current implementation.

```
with run_in_transaction():  # Will throw a WriteConflict error on __exit__
   user.username = "mcsimps2"
   user.save()
   with run_in_transaction():
        user.first_name = "Matt"
        user.save()
```

Personally, I've wrapped this functionality like so to take care of this issue.  I'm just not starting a new session if we're already in a transaction.

```
from mongoengine.context_managers import run_in_transaction as mongoengine_run_in_transaction

@contextmanager
def run_in_transaction(
    alias=DEFAULT_CONNECTION_NAME,
    session_kwargs=None,
    transaction_kwargs=None,
    disable_shared_sessions=False,
):
    session = _get_session()
    if not disable_shared_sessions and session and session.in_transaction:
        # Don't start a new transaction if we are already executing in one
        yield
    else:
        with mongoengine_run_in_transaction(
            alias=alias,
            session_kwargs=session_kwargs,
            transaction_kwargs=transaction_kwargs,
        ):
            yield
```

Conor94 commented:
Hi, are there any plans to release this feature? I noticed it wasn't included in the most recent release on Sep, 19, 2024: https://github.com/MongoEngine/mongoengine/releases/tag/v0.29.1
